### PR TITLE
move samples CRD to openshift api

### PR DESCRIPTION
**Pull request by gabemontero**
/assign @bparees 

@adambkaplan @openshift/openshift-team-developer-experience FYI

per discussion with @deads2k we don't have this under the `operator` dir because of what we did with our group name back in 4.0/4.1

/hold 

(I don't yet have the openshift/origin PR queued up per the doc in this repo's README)

also, I need to do something equivalent to https://github.com/openshift/api/pull/475 for our new dir, but wanted to get a sanity check before I went too far 

So far, these changes include generation/validations from 
- `make update`
- `make verify`

### CODE REVIEW

**Comment by bparees** [comment]
needs the yaml generated?

**Comment by bparees** [comment]
(seems sane otherwise)

File: samples/v1/types_config.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+	// facilitates the maintenance of this operator's ClusterOperator object.
+	MigrationInProgress ConfigConditionType = "MigrationInProgress"
+	// ImportImageErrorsExist registers any image import failures, separate from ImageChangeInProgress,
+	// so that we can a) indicate a problem to the ClusterOperator status, b) mark the current
+	// change cycle as complete in both ClusterOperator and Config; retry on import will
+	// occur by the next relist interval if it was an intermittent issue;
+	ImportImageErrorsExist ConfigConditionType = "ImportImageErrorsExist"
+	// numConfigConditionType is a helper constant that captures the number possible conditions
+	// defined above in this const block
+	numconfigConditionType = 7
```
Comment (adambkaplan) [inline samples/v1/types_config.go]: @gabemontero now that we've moved this to api, do we need it any more? It isn't exported, so I don't see what if anything could use it.

Comment (gabemontero) [reply samples/v1/types_config.go]: yep I should remove that thanks

**Comment by gabemontero** [comment]
got an assist from @benjaminapetersen !! 

@bparees @adambkaplan - he says he was told several times that the convention for the godoc in this case is that it should start with a lowercase character to match the json annotations on the struct fields

I'll be making that change shortly, unless you catch this and object beforehand

**Comment by bparees** [comment]
> @bparees @adambkaplan - he says he was told several times that the convention for the godoc in this case is that it should start with a lowercase character to match the json annotations on the struct fields

yes that sounds correct.

**Comment by gabemontero** [comment]
OK no more WIP elements that I'm aware of for this one @bparees

**Comment by gabemontero** [comment]
/assign @adambkaplan 

(for adding the lgtm if appropriate)

File: samples/v1/types_config.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+	// change cycle as complete in both ClusterOperator and Config; retry on import will
+	// occur by the next relist interval if it was an intermittent issue;
+	ImportImageErrorsExist ConfigConditionType = "ImportImageErrorsExist"
+)
+
+// ConfigCondition captures various conditions of the Config
+// as entries are processed.
+type ConfigCondition struct {
+	// Type of condition.
+	Type ConfigConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=ConfigConditionType"`
```
Comment (adambkaplan) [inline samples/v1/types_config.go]: @gabemontero godoc here needs to conform to conventions (lowercase, starts with name of the field).

Comment (gabemontero) [reply samples/v1/types_config.go]: udpate pushed @adambkaplan thanks

File: samples/v1/types_config.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Config contains the configuration and detailed condition status for the Samples Operator.
+type Config struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata" protobuf:"bytes,1,opt,name=metadata"`
+
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
```
Comment (adambkaplan) [inline samples/v1/types_config.go]: needs godoc - ex: `// spec is the desired configuration of the Samples Operator`

File: samples/v1/types_config.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+	// +optional
+	Status ConfigStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
```
Comment (adambkaplan) [inline samples/v1/types_config.go]: needs godoc: ex: `// status contains the observed configuration and other details describing the state of the Samples Operator.`

File: samples/v1/types_config.go (hunk 3 above)
Comment (gabemontero) [reply samples/v1/types_config.go]: yeah I went back and forth on this ... some of the existing examples in openshift/api seemed to eschew comments here in favor of the comments in the types below (i.e. ConfigSpec in this case).

see 
- https://github.com/openshift/api/blob/master/console/v1/types_console_cli_download.go
- https://github.com/openshift/api/blob/master/operator/v1/types_authentication.go
- https://github.com/openshift/api/blob/master/operator/v1/types_openshiftapiserver.go

and many other @adambkaplan 

I'd rather stay consistent with the pattern established elsewhere.

You OK with that ?

Comment (adambkaplan) [reply samples/v1/types_config.go]: There's plenty of other counter-examples. At this point if someone complains about the content in `oc explain` they can file a BZ.

Comment (gabemontero) [reply samples/v1/types_config.go]: True (which is why I went back and forth) ... in the end, looking at the generated CRD yaml, there are descriptions for ConfigSpec/ConfigStatus so I expect we will be OK

**Pull request merged**